_P221_LandisGyrUH550.ino

```diff
@@ -0,0 +1,450 @@
+      Device[deviceCount].PullUpOption = false;
+      Device[deviceCount].InverseLogicOption = false;
+      Device[deviceCount].FormulaOption = true;
+      Device[deviceCount].ValueCount = 1;
+      Device[deviceCount].SendDataOption = true;
+      Device[deviceCount].TimerOption = true;
+      Device[deviceCount].GlobalSyncOption = true;
+      break;
+    }
+
+    case PLUGIN_GET_DEVICENAME:
+    {
+      string = F(PLUGIN_NAME_221);
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+    {
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_221));
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      Plugin_221_init = true;
+
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_LOAD:
+    {
+      byte valueSet = Settings.TaskDevicePluginConfig[event->TaskIndex][1];
+
+      String optionsValueSet[7] = {  F("Quantity of Heat"),
+                                      F("Volume"),
+                                      F("Quantity of Heat previous Year"),
+                                      F("Volume previous Year"),
+                                      F("maxPower"),
+                                      F("maxInflowTemp"),
+                                      F("maxReflowTemp") };
+      addFormSelector(F("Valueset"), F("plugin_221_query"), 7, optionsValueSet, NULL, valueSet );
+
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SAVE:
+    {
+      Settings.TaskDevicePluginConfig[event->TaskIndex][0] = getFormItemInt(F("plugin_221"));
+      Settings.TaskDevicePluginConfig[event->TaskIndex][1] = getFormItemInt(F("plugin_221_query"));
+
+      Plugin_221_init = false; // Force device setup next time
+      success = true;
+      break;
+    }
+
+    case PLUGIN_READ:
+    {
+
+      if (Plugin_221_init)
+      {
+        String log = F("LUGUH: start read cycle: task <nr> : ");
+        log += event->TaskIndex;
+        log += " : "; 
+        log += getTimeString(':', true);
+        addLog(LOG_LEVEL_INFO, log);
+
+        switch(Settings.TaskDevicePluginConfig[event->TaskIndex][1])
+        {
+          case 0:
+          {
+            messageProcessingComplete = false;
+
+            PIN_SER_RX = Settings.TaskDevicePin1[event->TaskIndex];
+            PIN_SER_TX = Settings.TaskDevicePin2[event->TaskIndex];
+
+            ESPeasySoftwareSerial softSerial(PIN_SER_RX, PIN_SER_TX, false); // Initialize serial
+            char *index = 0;
+
+            pinMode(PIN_SER_RX, INPUT);
+            pinMode(PIN_SER_TX, OUTPUT);
+
+            //read Landis&Gyr
+            // 40 NUL chars + "/?!" + CR LF in 7 bit and EVEN parity
+            byte initstr[] = "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xAF\x3F\x21\x8D\x0A";
+
+            byte r = 0;
+            byte to = 0;
+            byte bufpos;
+            char message[255];

```

Comment:
This message is now allocated on cont stack (Arduino space) and we're really low on stack here.
So please consider the max. allowed return message.
It looks like the parsing is just looking for some string and then getting the next few bytes.
But also during parsing the message, some rather large arrays (2* 80 bytes + 15 + 6*70 + 2*80) are allocated on the cont stack.
So we're already at 1010+ bytes allocated on the stack during parsing and that's only the big arrays I saw when briefly looking at the code.
Keep in mind this stack is 4kB in size and when calling these functions the stack is already being used.